File: crates/normalize-languages/tests/fixtures/swift/variants.swift
// variants.swift is a completeness matrix: one small, commented construct per
// grammar-legal node-type variant found while applying
// docs/query-testing-methodology.md to swift.{tags,calls,imports,complexity,types}.scm.
// Every construct here is annotated with which field/variant it exercises,
// verified against real parse output via `normalize syntax query`/
// `normalize syntax ast` (node-types.json for arborium-swift 2.17.0), not
// assumed. A NEGATIVE section at the bottom holds constructs that must NOT
// match, so regressions in either direction are caught.

import Foundation
// import_declaration: dotted path (identifier node covers the whole
// "Foundation.NSObject" span, including the `.`, and the import-kind
// keyword doesn't interfere).
import class Foundation.NSObject

// --- tags.scm: function_declaration.name variants --------------------------

// plain_name — name: simple_identifier
func plainFunction() {}

infix operator +++: AdditionPrecedence

struct Vector {
    var x: Int
    var y: Int

    // custom_operator — name: custom_operator (multi-char operator overload)
    static func +++ (lhs: Vector, rhs: Vector) -> Vector {
        return Vector(x: lhs.x + rhs.x, y: lhs.y + rhs.y)
    }

    // standard-operator overload — name: literal operator token ("==").
    // Verified this is a REAL field child (`normalize syntax query
    // '(function_declaration name: "==" @name)'` matches), not the
    // return-type-mislabeling quirk documented in swift.tags.scm.
    static func == (lhs: Vector, rhs: Vector) -> Bool {
        return lhs.x == rhs.x && lhs.y == rhs.y
    }

    // compound-assignment operator overload — name: "+=" literal token.
    static func += (lhs: inout Vector, rhs: Vector) {
        lhs.x += rhs.x
        lhs.y += rhs.y
    }
}

// --- tags.scm: class_declaration.name variants ------------------------------

// plain_type_name — name: bare type_identifier
class PlainClass {}

// extension_name — name: user_type -> type_identifier. Distinct shape from
// PlainClass above; silently unhandled before this fix.
extension PlainClass {
    func extraMethod() {}
}

// extension with a `where` clause on a generic stdlib type — same
// user_type-wrapped name shape as the plain extension above.
extension Array where Element == Int {
    func sumAll() -> Int {
        return self.reduce(0, +)
    }
}

// --- tags.scm: enum_entry multi-name completeness ---------------------------

enum Status {
    // single case — name: simple_identifier
    case ready
    // associated-value case — name: simple_identifier (the payload type
    // doesn't change how `name` is tagged)
    case failed(String)
    // comma-separated multi-name case — name field is multiple=true and (per
    // real parse output) tags BOTH identifiers, unlike the analogous Go
    // const_spec bug from batch 1.
    case paused, cancelled
}

// --- tags.scm: member property let/var + local-variable exclusion ----------

class Holder {
    // member let — @definition.constant
    let readOnly: Int = 1
    // member var — @definition.var
    var mutable: Int = 2
    // computed property (still a property_declaration, still a direct child
    // of class_body) — @definition.var
    var computed: Int {
        return readOnly + mutable
    }

    func useLocals() {
        // NEGATIVE: local let/var inside a function body — same node kind
        // (property_declaration) as the member properties above, but must
        // NOT be captured as @definition.constant/@definition.var (the
        // ancestor-scoped class_body/enum_class_body restriction excludes
        // these; they are children of `statements`, not class_body).
        let localReadOnly = 5
        var localMutable = 6
        localMutable += localReadOnly
        print(localMutable)
    }
}

enum WithComputed {
    case a, b

    // computed property inside an enum — enum_class_body variant of the
    // class_body ancestor restriction above.
    var isA: Bool {
        return self == .a
    }
}

// --- tags.scm: protocol requirements ----------------------------------------

protocol Describable {
    // protocol_property_declaration — name: pattern -> simple_identifier
    var label: String { get }
    // protocol_function_declaration — name: simple_identifier
    func describe() -> String
    // associatedtype_declaration — name: type_identifier
    associatedtype Value
}

// --- calls.scm: call_expression.function variants ---------------------------

// plain_call — function: identifier
func identity(_ x: Int) -> Int {
    return x
}

struct Box {
    var value: Int
    // method_call — function: navigation_expression -> simple_identifier
    func get() -> Int {
        return value
    }
}

func callVariants() {
    _ = identity(1) // plain_call
    let b = Box(value: 5)
    _ = b.get() // method_call: navigation_expression, target: simple_identifier
    _ = Optional<String>(nil) // NOT relevant here; see constructor call below
}

// --- calls.scm: force-unwrap call (postfix_expression) ----------------------

final class Runner {
    var completion: (() -> Void)?

    func run() {
        // force-unwrap call — callee: postfix_expression(target:
        // simple_identifier, operation: bang). Previously unmatched.
        completion!()
        // optional-chaining call — callee: plain simple_identifier (the `?`
        // does not change call_expression's shape); already matched by the
        // plain_call pattern.
        completion?()
    }
}

// --- calls.scm: generic type instantiation call (constructor_expression) ---

struct GenericBox<T> {
    var value: T
}

func constructorVariants() {
    // NOT a call_expression — a distinct constructor_expression node with a
    // `constructed_type: (user_type (type_identifier))` field. Previously
    // unmatched by any calls.scm pattern.
    _ = GenericBox<Int>(value: 1)
    _ = Optional<String>(nil)
}

// --- complexity.scm: guard / switch_entry / conjunction / disjunction ------

func complexityVariants(_ n: Int, flag: Bool) -> String {
    // guard_statement — early-exit branch, previously uncounted.
    guard n > 0 else {
        return "non-positive"
    }
    // conjunction_expression (&&) / disjunction_expression (||) — previously
    // uncounted.
    if n > 0 && flag || n < 100 {
        return "complex"
    }
    switch n {
    case 1:
        return "one"
    case 2, 3:
        return "two-or-three"
    case let x where x > 10:
        return "big"
    default:
        return "other"
    }
}

// --- NEGATIVE cases: constructs that must NOT match -------------------------

func negativeCallVariants() {
    // NEGATIVE: function: call_expression (curried call) — the outer call
    // must not produce a @call/@name capture.
    _ = makeAdder()(1)

    // NEGATIVE: function: lambda_literal (immediately-invoked closure) — an
    // anonymous callee, must not produce a @call/@name capture.
    _ = { (x: Int) -> Int in x * 2 }(5)

    // NEGATIVE: function: array_type literal ([Int](...)) — a bracket
    // type-literal callee, not a declared symbol name, must not produce a
    // @call/@name capture.
    _ = [Int](repeating: 0, count: 3)
}

func makeAdder() -> (Int) -> Int {
    return { delta in delta }
}

func negativeLocalPropertyVariants() {
    // NEGATIVE: local let/var must never be tagged @definition.constant/
    // @definition.var (covered structurally by Holder.useLocals above; this
    // top-level-function copy exercises the same shape outside any class
    // body at all).
    let notAMember = 1
    var alsoNotAMember = 2
    alsoNotAMember += notAMember
    print(alsoNotAMember)
}

File: crates/normalize-languages/tests/fixtures/go/variants.go
// variants.go is a completeness matrix: one small, commented construct per
// grammar-legal node-type variant found while applying
// docs/query-testing-methodology.md to go.{tags,calls,imports,complexity,types}.scm.
// Every construct here is annotated with which field/variant it exercises,
// verified against real parse output via `normalize syntax query`/
// `normalize syntax ast` (node-types.json for arborium-go 2.17.0), not
// assumed. A NEGATIVE section at the bottom holds constructs that must
// NOT match, so regressions in either direction are caught.
package variants

import (
	"fmt"

	// import_spec.name: dot — dot import (blanket-imports strings' exports).
	. "strings"
	// import_spec.name: blank_identifier — side-effect-only import.
	_ "os"
	// import_spec.name: package_identifier — aliased import.
	f "fmt"
	// import_spec.path: raw_string_literal — grammar-legal but vanishingly
	// rare in practice (gofmt never emits it); this is the only realistic
	// way to exercise it.
	`errors`
)

// --- calls.scm / tags.scm: call_expression.function variants ---------------

// plain_call — function: identifier
func identity(x int) int {
	return x
}

// scoped_call — function: selector_expression, package-qualified
func sizeOf() int {
	return 0
}

type holder struct {
	value int
}

// method_call — function: selector_expression, method on a value
func (h holder) len() int {
	return h.value
}

func callVariants() {
	_ = identity(1)                  // plain_call
	_ = (identity)(1)                // parenthesized_expression wrapping identifier
	h := holder{value: 5}
	_ = h.len()                      // method_call: selector_expression
	_ = (h.len)()                    // parenthesized_expression wrapping selector_expression
	_ = fmt.Sprintf("%d", sizeOf())  // scoped_call: pkg.Func()
	_ = f.Sprintf("%d", 1)           // scoped_call via aliased import
	_ = Repeat("x", 2)               // dot-imported call: strings.Repeat via `. "strings"`
	err := errors.New("x")
	_ = err
}

// --- calls.scm write-context: @call vs @call.write --------------------------

func writeContextCall() {
	var result int
	result = identity(1)  // @call.write: assignment RHS
	result += identity(2) // @call.write: compound-assignment RHS
	_read := identity(3)  // @call: let-bound (read context, not write)
	_ = result
	_ = _read
}

// --- tags.scm/types.scm: type_spec vs type_alias ----------------------------

// type_spec — a named type definition (struct)
type Plain struct {
	X int
}

// type_alias — `type X = Y`, a distinct node type from type_spec (no `=`
// token appears in a type_spec). Silently unhandled before this fix.
type PlainAlias = Plain

// Generic type_spec: type_parameters is populated but the node is still
// type_spec, so this must be found the same way as Plain.
type Generic[T any] struct {
	Value T
}

// --- tags.scm/types.scm: qualified_type reference ---------------------------

type qualifiedRef struct {
	// qualified_type: package + name fields, both single-variant (fully
	// covered already — included here for completeness-matrix documentation).
	W fmt.Stringer
}

// --- tags.scm: const_spec multi-name completeness ---------------------------

const (
	// const_spec.name field: multiple=true, and per real parse output
	// (verified: only the FIRST identifier in a comma list is actually
	// tagged with the `name` field by tree-sitter-go; later names in the
	// same spec are unfielded children). A field-constrained pattern drops
	// every name after the first; a positional pattern catches all of them.
	SingleConst = 1
	MultiA, MultiB = 2, 3 // both MultiA and MultiB must be found
)

// --- calls.scm/tags.scm: deliberately-excluded call_expression.function ----
// variants (documented negative cases, not bugs — see go.calls.scm/
// go.tags.scm's own comments for the reasoning: none of these have a
// stable, nameable callee).

func higherOrder() func(int) int {
	return func(delta int) int {
		return delta
	}
}

var dispatch = []func(){
	func() { fmt.Println("zero") },
}

func negativeCallVariants() {
	// NEGATIVE: function: call_expression (curried call) — must not
	// produce any @call/@name capture for the outer call.
	_ = higherOrder()(1)

	// NEGATIVE: function: func_literal (immediately-invoked closure) —
	// must not produce any @call/@name capture. This is the idiomatic
	// goroutine/defer shape; excluded ONLY when called inline like this,
	// not when assigned to a name first (see dispatch above, called below).
	func() {
		fmt.Println("iife")
	}()

	// NEGATIVE: function: index_expression (dispatch-table call) — must
	// not produce any @call/@name capture.
	dispatch[0]()
}
